modules/flowable-engine/src/main/java/org/flowable/engine/impl/util/TaskHelper.java

```diff
@@ -59,41 +59,53 @@
  */
 public class TaskHelper {
 
-    public static void completeTask(TaskEntity taskEntity, Map<String, Object> variables,
-            Map<String, Object> transientVariables, boolean localScope, CommandContext commandContext) {
-
+    public static void completeTask(TaskEntity taskEntity, ScopedVariableContainerHelper scopedVariableContainerHelper,
+                                    CommandContext commandContext) {
         // Task complete logic
 
+        Map<String, Object> variables = null;
+
         if (taskEntity.getDelegationState() != null && taskEntity.getDelegationState() == DelegationState.PENDING) {
             throw new FlowableException("A delegated task cannot be completed, but should be resolved instead.");
         }
 
-        if (variables != null) {
-            if (localScope) {
-                taskEntity.setVariablesLocal(variables);
+        if (scopedVariableContainerHelper.hasAnyVariables()){
+            variables = scopedVariableContainerHelper.getAllVariables();
+        }
+
+        if (scopedVariableContainerHelper.hasAnyVariables()) {
+            if (scopedVariableContainerHelper.hasVariablesLocal()) {
+                taskEntity.setVariablesLocal(scopedVariableContainerHelper.getVariablesLocal());
 
+                if (scopedVariableContainerHelper.hasVariables()){
```

Comment:
Why id the hasVariables() check inside an hasVariablesLocal() check? Is it correct?

Reply:
Check it now please